contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.2.0--2.3.0.sql

```diff
@@ -2577,6 +2577,254 @@ AS 'babelfishpg_tsql', 'update_user_catalog_for_guest';
+  	      END
+  	    WHEN st.is_table_type = 1 THEN 0
+  	    ELSE sdit.ss_data_type
+  	  END
+  	AS sys.tinyint) AS SS_DATA_TYPE
+  , CAST(ss.proname AS sys.sysname) AS original_procedure_name
+  FROM 
+  ( 
+    -- CTE to query procedures related to bbf
+     WITH bbf_proc AS (
+   	   SELECT
+         p.proname as proname,
+         p.proargnames as proargnames,
+         p.proargmodes as proargmodes,
+         p.prokind as prokind,
+         p.proretset as proretset,
+         p.prorettype as prorettype,
+         p.pronamespace as pronamespace,
+         p.proallargtypes as proallargtypes,
+         p.proargtypes as proargtypes
+       FROM 
+         pg_proc p
+       WHERE (
+	     p.pronamespace in (select schema_id from sys.schemas union all select oid from pg_namespace where nspname = 'sys')
+	     AND (pg_has_role(p.proowner, 'USAGE') OR has_function_privilege(p.oid, 'EXECUTE'))
+	     AND (p.probin like '{%typmod_array%}'-- typmod array exists for those that are user-defined
+	       OR p.proname like 'sp\_%' -- filter out internal babelfish-specific procs
+	       OR p.proname like 'xp\_%'
+	       OR p.proname like 'dm\_%'
+	       OR p.proname like 'fn\_%'))
+	)
+	       
+     SELECT -- Selects all parameters (input and output), but NOT return values
+       p.proname as proname,
+       p.proargnames as proargnames,
+       p.proargmodes as proargmodes,
+       p.prokind as prokind,
+       p.proretset as proretset,
+       p.prorettype as prorettype,
+       p.pronamespace as pronamespace,
+       information_schema._pg_expandarray(
+       COALESCE(p.proallargtypes,
+         CASE 
+           WHEN p.prokind = 'f' THEN (CAST(p.proargtypes AS oid[]))
+           ELSE CAST(p.proargtypes AS oid[])
+         END
+       )) AS x
+    FROM bbf_proc p
+      
+    UNION ALL
+      
+    SELECT -- Selects all return values
+      p.proname as proname,
+      p.proargnames as proargnames,
+      p.proargmodes as proargmodes,
+      p.prokind as prokind,
+      p.proretset as proretset,
+      p.prorettype as prorettype,
+      p.pronamespace as pronamespace,
+      NULL AS x -- null value indicates that we are retrieving the return values of the proc/func
+    FROM bbf_proc p
+  ) ss
+  INNER JOIN pg_catalog.pg_namespace ns ON ns.oid = ss.pronamespace
+  LEFT JOIN sys.babelfish_namespace_ext bne ON ns.nspname = bne.nspname 
+  LEFT JOIN sys.databases d ON d.database_id = bne.dbid
+  LEFT JOIN sys.types st ON ((ss.x IS NOT NULL AND st.user_type_id = (ss.x).x) OR (ss.x IS NULL AND st.user_type_id = ss.prorettype))
+  -- Because spt_datatype_info_table does contain user-defind types and their names,
+  -- the join below allows us to retrieve the system type name on what the user-defined type was based on.
+  LEFT JOIN sys.spt_datatype_info_table sdit ON sdit.type_name = (SELECT name FROM sys.types WHERE user_type_id = st.system_type_id)
+WHERE (ss.proargmodes[(ss.x).n] in ('i', 'o', 'b')  OR ss.proargmodes[(ss.x).n] is NULL);
```

Comment:
I feel that the first disjunction term could be moved to line 2802 and the second one could be removed completely

Reply:
1. `ss.proargmodes[(ss.x).n] in ('i', 'o', 'b')`
I don't think this one can be moved inside 2802 because we can only get the `.n` value from `(ss.x)` after the projection. Before the projection from the inner query, we only have one row per function/procedures. 

2. `ss.proargmodes[(ss.x).n] is NULL` 

This was a mistake and was actually supposed to be two cases (`ss.proargmodes is NULL OR ss.x is NULL`). I accidentally left this in and it seemed to have worked out equivalently to those (sheer luck) but will change that in the next revision. We do need this though because

- `ss.proargmodes is NULL` is needed since in a procedure, if all parameters are inputs, it will be NULL (defined in the pg_proc documentation). 
- `ss.x is NULL` we need this because of the return_values query in the inner query (2805-2813).